### ISSUE TITLE
"kubectl explain" should be able to explain "apiservices" and "customresourcedefinition"

### ISSUE THREAD

**Issue by xiangpengzhao** [kubernetes/kubernetes#49465]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
```
# kubectl explain apiservices
error: group apiregistration.k8s.io has not been registered
```

```
# kubectl explain customresourcedefinition
error: group apiextensions.k8s.io has not been registered
```

**What you expected to happen**:
`kubectl explain` should be able to explain the above two resources. 

Print something as below, etc.
```
DESCRIPTION:
APIService represents a server for a particular GroupVersion. Name must be "version.group".

FIELDS:
......
```

**How to reproduce it (as minimally and precisely as possible)**:
Run the commands above.

**Anything else we need to know?**:
/sig cli
/sig api-machinery

**Environment**:
- Kubernetes version (use `kubectl version`): a9bf44101b3ec13a3e94df926c95ba5384d18e28
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by xiangpengzhao** [comment]
@kubernetes/sig-cli-bugs

**Comment by caesarxuchao** [comment]
cc @mbohlool

**Comment by mbohlool** [comment]
This look like a problem with old swagger 1.2. We do not aggregate swagger 1.2 spec. When kubectl move to OpenAPI, it should by fix this by design.

**Comment by nikhita** [comment]
This is still not fixed (even after kubectl moving to OpenAPI).

There are two problems here:

1. kubectl uses groups registered into a group registry. This can be fixed by:

```diff
--- a/pkg/kubectl/cmd/explain.go
+++ b/pkg/kubectl/cmd/explain.go
@@ -26,7 +26,6 @@ import (
 	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
 	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
 	"k8s.io/kubernetes/pkg/kubectl/explain"
-	"k8s.io/kubernetes/pkg/kubectl/scheme"
 	"k8s.io/kubernetes/pkg/kubectl/util/i18n"
 )
 
@@ -91,7 +90,6 @@ func RunExplain(f cmdutil.Factory, out, cmdErr io.Writer, cmd *cobra.Command, ar
 		return err
 	}
 
-	// TODO: We should deduce the group for a resource by discovering the supported resources at server.
 	fullySpecifiedGVR, groupResource := schema.ParseResourceArg(inModel)
 	gvk := schema.GroupVersionKind{}
 	if fullySpecifiedGVR != nil {
@@ -104,20 +102,12 @@ func RunExplain(f cmdutil.Factory, out, cmdErr io.Writer, cmd *cobra.Command, ar
 		}
 	}
 
-	if len(apiVersionString) == 0 {
-		groupMeta, err := scheme.Registry.Group(gvk.Group)
-		if err != nil {
-			return err
-		}
-		apiVersion = groupMeta.GroupVersion
-
-	} else {
-		apiVersion, err = schema.ParseGroupVersion(apiVersionString)
-		if err != nil {
+	if len(apiVersionString) != 0 {
+		if apiVersion, err = schema.ParseGroupVersion(apiVersionString); err != nil {
 			return err
 		}
+		gvk = apiVersion.WithKind(gvk.Kind)
 	}
-	gvk = apiVersion.WithKind(gvk.Kind)
 
 	resources, err := f.OpenAPISchema()
 	if err != nil {
```

However, after this it gives:

```
error: Couldn't find resource for "apiextensions.k8s.io/v1beta1, Kind=CustomResourceDefinition"
```

2.  This occurs because of incorrect openAPI schema. The schema does not contain apiextensions and apiregisteration.

kubectl explain looks up the gvk of the resource in question from a list of resources:

https://github.com/kubernetes/kubernetes/blob/ff626a35fe35f6d219a3f95b39df7afd4cab9b6b/pkg/kubectl/cmd/util/openapi/openapi.go#L71-L77

It gets the gvk here:

https://github.com/kubernetes/kubernetes/blob/ff626a35fe35f6d219a3f95b39df7afd4cab9b6b/pkg/kubectl/cmd/util/openapi/openapi.go#L59

`parseGroupVersionKind` finds the gvk through `groupVersionKindExtensionKey`:

https://github.com/kubernetes/kubernetes/blob/ff626a35fe35f6d219a3f95b39df7afd4cab9b6b/pkg/kubectl/cmd/util/openapi/openapi.go#L83-L87

Here, `groupVersionKindExtensionKey` is `"x-kubernetes-group-version-kind"` and is the key used to lookup the GroupVersionKind value for an object definition from the definition's "extensions" map.

However, we do not generate `x-kubernetes-` extensions for apiextensions-apiserver and aggregator yet (https://github.com/kubernetes/kubernetes/issues/52741), due to which it fails.

/reopen

**Comment by nikhita** [comment]
Looks like the blocker was fixed in https://github.com/kubernetes/kubernetes/pull/64174 🎉 

I will move forward with the fix.

**Comment by liggitt** [comment]
fixed in 1.11 in https://github.com/kubernetes/kubernetes/pull/64174
